### PR TITLE
Refactor distribution hierarchy

### PR DESCRIPTION
**Pull request by marcelluethi**
This PR refactors the distribution hierarchy. In particular it adds a common trait that all distributions share. 
This requires IndependenDistributions to sum the logProbs to satisfy the common interface. To enable other operations that summing probabilities, an method `lementwiseLogProbs  is added.  

This PR also makes several smaller fixes, improving type safety and handling of corner cases. 

Finally, it also exposes Random and some more stats operations on package level. 

Note, this PR builds on the currently unmerged PR #64 . Once this is merged it might become much easier to review.

### CODE REVIEW

File: core/src/main/scala/dimwit/stats/MultivariateDistributions.scala [hunk 1]
```diff
@@ -43,33 +41,31 @@ class Dirichlet[L: Label](
...
+    // Sample from categorical n times using splitvmap, then bincount
+    trait Draws derives Label
+    val draws = key.splitvmap(Axis[Draws] -> n)(k => categorical.sample(k))
     Tensor.fromPy(VType[Int])(
-      Jax.jnp.bincount(samples, length = probs.shape.dimensions(0))
+      Jax.jnp.bincount(draws.jaxValue, length = probs.shape.dimensions(0))
     )
 
-class Categorical[L: Label](val probs: Tensor1[L, Float]) extends MultivariateDistribution[EmptyTuple, Int]:
+class Categorical[L: Label](val probs: Tensor1[L, Prob]) extends UnivariateDistribution[Int]:
```
[the remark this answers is not in the document]
Comment (benikm91) [reply core/src/main/scala/dimwit/stats/MultivariateDistributions.scala]: If we add, don't forget to add to export

**Review by benikm91 — APPROVED** [review]
A comment suggests a minor improvement. Yet, changes are overall good and can be merged.

**Comment by benikm91** [comment]
Two questions arose while rebaseing my GPT-train branch to this branch:

1. Is the Float upper bound still necessary?
```scala
opaque type LogProb <: Float = Float
opaque type Prob <: Float = Float
```

2. Should softmax be changed to return "Prob", now that Random and Distribution are part of the core
```scala
def softmax[L: Label](t: Tensor1[L, Float]): Tensor1[L, Prob] =
  Tensor(Jax.jnn.softmax(t.jaxValue, axis = 0))
```
Motivation:
```scala
// In sampling, I would need an explicit Prob cast, now:
Categorical(softmax(nextTokenLogits /! temperatur)).sample(sampleKey).item
```

**Comment by benikm91** [comment]
- Thanks for adding separate file
- Agree.
- Regarding softmax I am unsure. Let's keep it as it is in this PR.
- The upper bound just tells the outside that e.g. LogProb is a Float, so the whole Float API is part of the type definition. For tensor operations we already have the more restrictive IsFloating evidence, we don't need <: Float. I think, I added this long time ago, when I was still figuring out opaque types. => I removed it
